src/fiber_util/fiber_util.mli

```diff
@@ -62,3 +62,87 @@ module Cancellation : sig
     -> on_cancellation:(unit -> 'b Fiber.t)
     -> ('a * 'b outcome) Fiber.t
 end
+
+(** A observer for a value that's changing over time *)
+module Observer : sig
+  (** An observer is something that let's one monitor a value that is changing
+      over time.
+
+      We call a value that is changing over time and can be observed an
+      "observable". An "observer" is a routine watching this observable and
+      getting notified when the value change. An observable may have several
+      observers.
+
+      The goal of an observer is to be up-to-date with the observable. It is not
+      guaranteed that the observer will observe all possible intermediate values
+      of the observable. For instance, if an observable changes several time in
+      a row, by the time the observer is awoken it will receive the latest value
+      of the observable. {!await} will only suspend the current fiber if the
+      current value of the observable is the same as the last one that was
+      reported.
+
+      Given that an observer is stateful, it shouldn't be shared between
+      different consumers. Each consumer that wants to watch an observable
+      should create it's wn observer. Sharing an observer would lead to
+      consumers not being up-to-date with the latest value of the observable. An
+      observer should always be unsubscribed to via {!unsubscribe} when done
+      with, otherwise the observable will continue updating it.
+
+      Finally, sometimes the value is big but only changes a little at a time.
+      For instance, an observable might represent "the set of current errors",
+      and the producer will only add a remove one error at time. To avoid
+      dumping a massive value on each update, we might instead want to send
+      small "diffs". In such a case, values reported by {!await} would be diffs
+      rather than full values and it is up to the consumer to reconstruct the
+      full value if they want it. Successive diffs would still be squashed so
+      that a call to {!await} returns them all at once. *)
+  type 'a t
+
+  (** [await t] returns the currently observed value. If the value hasn't been
+      updated since the last [await t] call, this call will block. If the
+      observable is closed, or unsubscribe is called this will return [None].
+      After [None], all subsequent calls will return [None] immediately *)
+  val await : 'a t -> 'a option Fiber.t
```

Comment:
The name `await` seems both vague and confusing because of the common `async`/`await` idiom.

How about `observe`? It's an observer after all, so observing is what it does.